felix/wireguard/bootstrap.go

```diff
@@ -17,110 +17,401 @@ package wireguard
 	nodeName := configParams.FelixHostname
 
-	_, dbgBootstrap := os.LookupEnv("FELIX_DBG_WGBOOTSTRAP")
+	logCxt := log.WithFields(log.Fields{
+		"iface":    wgDeviceName,
+		"nodeName": nodeName,
+	})
+	logCxt.Debug("Bootstrapping wireguard")
+
+	if !configParams.WireguardEnabled || configParams.WireguardInterfaceName == "" {
+		// Always remove wireguard configuration if not enabled.
+		logCxt.Info("Wireguard is not enabled - ensure no wireguard config")
+		return nil, removeWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
+		// This should not change for a given deployment. We only need to handle bootstrapping for clusters where
+		// host encryption is enabled.
+		logCxt.Debug("Host encryption is not enabled - no wireguard bootstrapping required")
+		return nil, nil
+	}
+
+	// Get the local public key and the peer public keys currently programmed in the kernel.
+	kernelPublicKey, kernelPeerKeys := getWireguardInfo(logCxt, wgDeviceName, getWireguardHandle)
+
+	// If there is no valid wireguard configuration in the kernel then remove all traces of wireguard.
+	if kernelPublicKey == "" || kernelPeerKeys.Len() == 0 {
+		logCxt.Info("No valid wireguard kernel routing - removing wireguard configuration completely")
+		return nil, removeWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	storedPublicKey, err := getPublicKeyForNode(logCxt, nodeName, calicoClient)
+	if err != nil {
+		// Could not determine the public key for this node.
+		return nil, fmt.Errorf("couldn't determine current wireguard configuration after %d retries: %v", bootstrapMaxRetries, err)
+	}
+
+	if storedPublicKey != kernelPublicKey {
+		// The public key configured in the kernel differs from the value stored in the node. Remove all wireguard
+		// configuration.
+		logCxt.Info("Found mismatch between kernel and datastore wireguard keys - removing wireguard configuration")
+		return nil, removeWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+	}
+
+	// The configured and stored wireguard key match.
+	logCxt.WithField("peerKeys", kernelPeerKeys).Info("Wireguard public key matches kernel")
+	return kernelPeerKeys, nil
+}
+
+// FilterTyphaEndpoints filters the supplied set of typha endpoints to the set where wireguard routing is most likely
+// to succeed. Any errors encountered are swallowed and the associated peer is just included.
+func FilterTyphaEndpoints(
+	configParams *config.Config,
+	v3Client clientv3.Interface,
+	typhas []discovery.Typha,
+	peersToValidate set.Set,
+) []discovery.Typha {
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
```

Comment:
Does this comment actually matter? configParams is the merged env var / datastore / config file value.

I don't think it matters, but probably not a good idea to assume that this is only ever set via an env var. For one, the operator is may change that behavior in the very near future (we're discussing some options here). And secondly, there are non-operator deployments that exist

Reply:
Well it matters in the sense that it's being used to determine when to tear down wireguard configuration. We don't really want to be doing that if we can help it and so we don't want to be doing it if your cluster should never be using host encryption or if you never enabled host encryption..... maybe it's more a docs thing.... If you want to turn off host encryption leave it enabled, disable fireguard, then disable host encryption then re-enable wireguard.  Yuck.

Reply:
I suppose I'd be inclined to nuke wireguard state whenever wireguard is disabled too.  If user has set wireguard to disabled then the end result we want is no wireguard device, no key in datastore etc.  Doing that tear-down pro-actively here means that we'd cover the case where the user disables both host encryption and wireguard in one config change.